feat: allow running single test (#3619)
scalameta/metals: docs/integrations/test-explorer.md

---
id: test-explorer
title: Test Explorer
---

Test Explorer is under active development, this page explain how it works under the hood.

> ## Test Discovery

Reviewer: One other thing that might be helpful to specify in here. For a client to use this, do they just set `testExplorerProvider` to true and then call it a day? Which approach that you outline is the default? Or will they both work by default?

Author: I've added a paragraph that should address these concerns, will it be good enough?

Reviewer: Hey! Yup, that's helpful, thanks!

---
id: test-explorer
title: Test Explorer
---

> Test Explorer is being under active development, this page explain how it works under the hood.

Reviewer: ```suggestion
Test Explorer is under active development, this page explain how it works under the hood.
```

## Test Discovery

Test Explorer provides two ways to interact with it:

> - first one is based on notifications being sent to the client and doesn't require from client to start any actions. All necessary updates are being immediately sent to it.

Reviewer: ```suggestion
- The first one is based on notifications being sent from Metals to the client and doesn't require the client to do any actions except listen for the notifications. All necessary updates are being immediately sent to it.
```

![](../diagrams/out/test-explorer-server-notifications.png)

> - client can also manually ask for test suites from the whole workspace, or for test cases for a given file.

Reviewer: ```suggestion
- The client can also manually ask Metals for test suites for the whole workspace, or for the test cases for a given file.
```

![](../diagrams/out/test-explorer-client-queries.png)

## How it works under the hood

> Test Explorer uses BSP's [debug request](https://build-server-protocol.github.io/docs/specification.html#debug-request) to run and debug test classes. TE reuses already available debug support in Metals and build servers such as Bloop or SBT.

Reviewer: ```suggestion
Test Explorer uses BSP's [debug request](https://build-server-protocol.github.io/docs/specification.html#debug-request) to run and debug test classes. TE reuses already available debug support in Metals and build servers such as Bloop or SBT.
```
I think this is the first place you use TE so you should do Test Explorer (TE) and then use TE form there on out, or just use Test Explorer.

Both run and debug under the hood use BSP's debug request.

### Running and debugging

> Following diagram ([source](https://github.com/scalacenter/bloop/blob/12bdc7f97cc3970d3e22a8b513f4b609c813f0a7/docs/assets/dap-example-metals.png)) represents what communication between language client, language server and build server. (It's very likely that in the future Debug Adapter will be inlined in the Metals to reduce number of moving parts).

Reviewer: > in the future Debug Adapter will be inlined in the Metals to reduce number of moving parts.

Really?

Author: Hmm, people on the streets talk too much :P

![](https://raw.githubusercontent.com/scalacenter/bloop/master/docs/assets/dap-example-metals.png)

> Current implementation of TE is a better version of code lenses, with more user friendly UX provided by the VSCode.

Reviewer: ```suggestion
The current implementation of TE is a better version of code lenses, with more user friendly UX provided by the VSCode.
```

### Getting and analyzing results

After test run was finished TE analyzes execution results and sets proper statutes for each test included in run.
![](https://imgur.com/vop095s.png)
![](https://imgur.com/TLhwEtG.png)

Such detailed information is available thanks to the [sbt-test-interface](https://github.com/sbt/test-interface) which abstracts over test framework internals, most important classes are: [Event](https://github.com/sbt/test-interface/blob/17a94641942546e21f4c6b300a3360be2d2888f6/src/main/java/sbt/testing/Event.java) and [Status](https://github.com/sbt/test-interface/blob/17a94641942546e21f4c6b300a3360be2d2888f6/src/main/java/sbt/testing/Status.java) which are used in the [sbt.ForkMain](https://github.com/sbt/sbt/blob/50c3cf10825bae3e89af11b3a2c9741e6d98add5/testing/agent/src/main/java/sbt/ForkMain.java). Almost all of build servers (if not all of them) uses `ForkMain` and `sbt-test-interface` to offload testing.

> For each test, there is fired an [event](https://github.com/sbt/sbt/blob/50c3cf10825bae3e89af11b3a2c9741e6d98add5/testing/agent/src/main/java/sbt/ForkMain.java#L394-L440) fired which contains all necessary information about execution: fully qualified name of test, execution status and duration.

Reviewer: ```suggestion
For each test, there is an [event](https://github.com/sbt/sbt/blob/50c3cf10825bae3e89af11b3a2c9741e6d98add5/testing/agent/src/main/java/sbt/ForkMain.java#L394-L440) fired which contains all necessary information about execution: fully qualified name of test, execution status and duration.
```
